### ISSUE TITLE
Organisation Thoughts

### ISSUE THREAD

**Issue by tpunt** [Sobak/php7-migration-guide#4]
The way the content is organised at the moment means that new classes have to awkwardly fit into the new functions section, while new interfaces have to be placed elsewhere altogether (in the new features section?). I think that restructuring the guide would be a good idea, as well as organising all of the content into the new structure here since some things will not cleanly fit into a single section.

The current structure is as follows:
- Backward incompatible changes
- New features
- Deprecated features in PHP 7.0.x
- Changed functions
- New functions
- Other changes to extensions
- Changes to the error reporting

A new potential structure may be (sections should probably be re-ordered):
- Backward Incompatible Changes
- New Features
- Deprecated Features in PHP 7.0.x
- Functions Changes
- New Functions and Methods
- New Classes and Interfaces
- Language Changes
- Extension Changes
- SAPI Changes
- Error Reporting Changes

Organising the content into this new structure may look as follows:
- Backward Incompatible Changes
  - Invalid octal integer literals are now parse errors
  - ...
- New Features
  - Combined Comparison Operator
  - Null Coalesce Operator
  - Scalar Type Declarations
  - Return Type Declarations
  - Unicode Codepoint Escape Syntax
  - Group use Declarations
  - Generator Return Expressions
  - Generator Delegation
- Deprecated Features in PHP 7.0.x
  - PHP 4-Style Constructors
  - Salt Option for password_hash()
- Functions Changes
  - Filtered unserialize()
  - session_start() Options
  - Improved assertions using assert() (Expectations)
- New Functions and Methods
  - Closure call() Method
  - Integer Division with intdiv()
  - preg_replace_callback_array() Function
  - error_clear_last() Function
- New Classes and Interfaces
  - IntlChar class
  - Exception hierarchy (*Error classes and Throwable interface)
- Language Changes
  - Fixes to foreach()'s Behaviour
  - Fixes to Custom Session Handler Return Values
  - Changes to list()'s Behaviour
  - Changes to Division by Zero Semantics
  - Integer Semantics
  - ZPP Failure on Overflow
  - Loosening Reserved Word Restrictions
  - Uniform Variable Syntax
  - Removal of Hex Support in Numerical Strings
  - Removal of Alternative PHP Tags
  - Removal of Multiple Default Blocks in Switch Statements
  - Removal of Deprecated Functionality
  - Removal of date.timezone Warning
- Extension Changes
  - JSON Extension Replaced with JSOND
- SAPI Changes
  - Removal of Dead Server APIs
- Error Reporting Changes
  - Exceptions in the Engine
  - Reclassification and Removal of E_STRICT Notices

That's roughly mapping my PHP 7 reference article into these sections (I'll update it with the missed features/changes tomorrow). Any thoughts on this?

**Comment by Sobak** [comment]
I'm fine with changing the structure. Current one is the slightly modified migration guide for 5.x version and as PHP 7 has way more changes, we can and probably should adjust the structure.

I have a few particular questions like
- where would you place removal of extensions like `ext/mysql`? In _Extension Changes_, right?
- removal of date.timezone warning is not a language change IMO (it should rather belong to backwards-incompatible changes)
- assertion changes are way more than changing function - ideally, it would belong to _Language Changes_ as well, but yes, we need to choose one

but those are details to discuss and it doesn't change the fact that your proposal is better than the current state. Let's see what @salathe thinks - he's the PHP Documentation Editor.

Thomas, thank you for your work on such detailed PHP 7 Reference and official migration guide!

**Comment by salathe** [comment]
The "new potential structure" pretty closely matches what is available in the [UPGRADING](https://github.com/php/php-src/blob/master/UPGRADING) file already. I would be inclined to go with the structure defined in that file.

**Comment by tpunt** [comment]
I'd be happy with following the same structure as in UPGRADING.

**Comment by salathe** [comment]
Any updates on this, guys? I'd be inclined to start moving this work to the documentation proper, as we're quickly progressing through the beta releases of PHP 7.

**Comment by tpunt** [comment]
I've merged all features from the PHP 7 reference into my fork of the migration guide (https://github.com/tpunt/php7-migration-guide/commit/9634878e90477af9924078ac7039fc2500a7c31b, parts may need to be reworded, examples cleaned up, etc). I've also started restructuring the guide to follow the one in UPGRADING (though haven't committed this to my fork yet). I hope to work more on this over the coming days.
